internal/oci/runtime_oci.go

```diff
@@ -635,14 +635,42 @@ func waitContainerStop(ctx context.Context, c *Container, timeout time.Duration,
 		if err := kill(pid); err != nil {
 			return fmt.Errorf("failed to kill process: %v", err)
 		}
+	case newTimeout := <-c.stopTimeoutChan:
+		// if a new timeout comes in,
+		// interrupt the old one, and start a new one
+		close(chControl)
+		return waitContainerStop(ctx, c, newTimeout, ignoreKill)
 	}
-
 	c.state.Finished = time.Now()
 	return nil
 }
 
 // StopContainer stops a container. Timeout is given in seconds.
 func (r *runtimeOCI) StopContainer(ctx context.Context, c *Container, timeout int64) error {
+	// first, need to check if the container is already stopping
+	c.stopLock.Lock()
+	if c.timesStopped > 0 {
+		// If so, we shouldn't wait forever on the opLock.
+		// This can cause issues where the container stop gets DOSed by a very long
+		// timeout, followed by someone sending a very short one.
+		// Instead, interrupt the other stop with this new one.
+		c.stopTimeoutChan <- time.Duration(timeout) * time.Second
```

Comment:
Isn't there a race if `waitContainerStop` is done with the select loop? This write would then block.

Reply:
who would have thought, such a simple solution is not that simple after all...

Reply:
okay I think I fixed this race by introducing a new channel.